flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java

```diff
@@ -0,0 +1,298 @@
+	}
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(JdbcBatchingOutputFormat.class);
+
+	private final JdbcExecutionOptions executionOptions;
+	private final ExecutorCreator<JdbcExec> statementRunnerCreator;
+	final RecordExtractor<In, JdbcIn> jdbcRecordExtractor;
+
+	private transient JdbcExec jdbcStatementExecutor;
+	private transient int batchCount = 0;
+	private transient volatile boolean closed = false;
+
+	private transient ScheduledExecutorService scheduler;
+	private transient ScheduledFuture<?> scheduledFuture;
+	private transient volatile Exception flushException;
+
+	JdbcBatchingOutputFormat(
+			JdbcConnectionProvider connectionProvider,
+			JdbcExecutionOptions executionOptions,
+			ExecutorCreator<JdbcExec> statementExecutorCreator,
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
+		this.statementRunnerCreator = statementExecutorCreator;
+		this.jdbcRecordExtractor = recordExtractor;
+	}
+
+	/**
+	 * Connects to the target database and initializes the prepared statement.
+	 *
+	 * @param taskNumber The number of the parallel instance.
+	 */
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		super.open(taskNumber, numTasks);
+		jdbcStatementExecutor = statementRunnerCreator.apply(getRuntimeContext());
+		try {
+			jdbcStatementExecutor.open(connection);
+		} catch (SQLException e) {
+			throw new IOException("unable to open JDBC writer", e);
+		}
+		if (executionOptions.getBatchIntervalMs() != 0 && executionOptions.getBatchSize() != 1) {
+			this.scheduler = Executors.newScheduledThreadPool(1, new ExecutorThreadFactory("jdbc-upsert-output-format"));
+			this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(() -> {
+				synchronized (JdbcBatchingOutputFormat.this) {
+					if (!closed) {
+						try {
+							flush();
+						} catch (Exception e) {
+							flushException = e;
+						}
+					}
+				}
+			}, executionOptions.getBatchIntervalMs(), executionOptions.getBatchIntervalMs(), TimeUnit.MILLISECONDS);
+		}
+	}
+
+	private void checkFlushException() {
+		if (flushException != null) {
+			throw new RuntimeException("Writing records to JDBC failed.", flushException);
+		}
+	}
+
+	@Override
+	public final synchronized void writeRecord(In record) {
+		checkFlushException();
+
+		try {
+			doWriteRecord(record);
+			batchCount++;
+			if (batchCount >= executionOptions.getBatchSize()) {
+				flush();
+			}
+		} catch (Exception e) {
+			throw new RuntimeException("Writing records to JDBC failed.", e);
+		}
+	}
+
+	void doWriteRecord(In record) throws SQLException {
+		jdbcStatementExecutor.process(jdbcRecordExtractor.apply(record));
+	}
+
+	@Override
+	public synchronized void flush() throws IOException {
+		checkFlushException();
+
+		for (int i = 1; i <= executionOptions.getMaxRetries(); i++) {
+			try {
+				attemptFlush();
```

Comment:
Here shouldn't we restart the timer? It seems like we may be in the situation where we flushed this batch because we reached the max size, so for the new one we have to set a new timer (renew the `BatchIntervalMs`), right?

Reply:
Yes. I would also replace `Thread.sleep(1000 * i);` with just rescheduling (retaining blocking semantics of `write`).
But I'm not sure if such changes should be in this PR.
WDYT?

Reply:
I agree that this seems to be more of a bug that deserves its own PR. Would you like to open a JIRA for this and assign it to you?

Reply:
Yes, that's a good idea.
Created a ticket for that: https://issues.apache.org/jira/browse/FLINK-16328